nemo/collections/tts/models/fastpitch_e2e.py

```diff
@@ -0,0 +1,873 @@
+            pitch = batch_dict.get("pitch", None)
+            energy = batch_dict.get("energy", None)
+            speaker = batch_dict.get("speaker_id", None)
+            durs = batch_dict.get("durations", None)
+        else:
+            audio, audio_lens, text, text_lens, attn_prior, pitch, durs, speaker = batch
+
+        mels, spec_len = self.fp_model.preprocessor(input_signal=audio, length=audio_lens)
+
+        (
+            mels_pred,
+            dec_lens,
+            durs_predicted,
+            log_durs_pred,
+            pitch_pred,
+            attn_soft,
+            attn_logprob,
+            attn_hard,
+            attn_hard_dur,
+            pitch,
+            energy_pred,
+            energy_tgt,
+        ) = self.fp_model(
+            text=text,
+            durs=durs,
+            pitch=pitch,
+            energy=energy,
+            speaker=speaker,
+            pace=1.0,
+            spec=mels if self.learn_alignment else None,
+            attn_prior=attn_prior,
+            mel_lens=spec_len,
+            input_lens=text_lens,
+        )
+        if durs is None:
+            durs = attn_hard_dur
+
+        audio = audio.unsqueeze(1)
+
+        fp_inter_mel = self.mel_loss_fn(spect_predicted=mels_pred, spect_tgt=mels)
+
+        # Skip batch if it is too short
+        if self.cfg.use_patches:
+            ids_str_max = dec_lens - self._cfg.segment_size // self._cfg.n_window_stride + 1
+            if len(ids_str_max[ids_str_max < 0]) > 0:
+                print('Encountered short batch')
+                return
+
+            mels_pred, ids_slice = rand_slice_segments(
+                mels_pred, dec_lens, self._cfg.segment_size // self._cfg.n_window_stride
+            )
+            audio = slice_segments(audio, ids_slice * self.cfg.n_window_stride, self._cfg.segment_size)
+            mels = slice_segments(mels, ids_slice, self._cfg.segment_size // self._cfg.n_window_stride)
+
+        audio_pred = self.hfg_model.generator(x=mels_pred)
+
+        audio_pred_mel, _ = self.fp_model.preprocessor(input_signal=audio_pred.squeeze(1), length=audio_lens)
+
+        optim_g, optim_d = self.optimizers()
+
+        # Train discriminator
+        optim_d.zero_grad()
+
+        mpd_score_real, mpd_score_gen, _, _ = self.hfg_model.mpd(y=audio, y_hat=audio_pred.detach())
+        loss_disc_mpd, _, _ = self.discriminator_loss(
+            disc_real_outputs=mpd_score_real, disc_generated_outputs=mpd_score_gen
+        )
+        msd_score_real, msd_score_gen, _, _ = self.hfg_model.msd(y=audio, y_hat=audio_pred.detach())
+        loss_disc_msd, _, _ = self.discriminator_loss(
+            disc_real_outputs=msd_score_real, disc_generated_outputs=msd_score_gen
+        )
+        loss_d = loss_disc_msd + loss_disc_mpd
+
+        self.manual_backward(loss_d)
+        norm_d = clip_grad_value_(
+            itertools.chain(self.hfg_model.msd.parameters(), self.hfg_model.mpd.parameters()), 20
```

Comment:
Can we use the `trainer.gradient_clip_val` value we usually have in our config instead of hardcoding to 20?

Reply:
1. We can't, since the code is written with automatic_optimization=False. because it is a GAN model. 
2. I think this is a redundant clipping, was temporary solution which gave 0 effect

Reply:
I meant that the model class could access the value `self.trainer.gradient_clip_val` directly (or get it elsewhere in the config). But if the clip value is not necessary, then it does not matter.